### ISSUE TITLE
add support for mqtt

### ISSUE THREAD

**Issue by MatzeJoerling** [ickerwx/tcpproxy#14]
Hi,
i have some envertech solar stuff and they send data to china cloud. i am intercepting them with your tcppproxy, no encryption, easy binary protocol. I want to have them (data Power, Voltage ...) in my node-red dashboard. Easiest way would be you tcpproxy send all server responses to mqtt topic. 
So, it would be  an -om module... needs mqtt server, user pass (optional) and topic. Whole Data HEX encoded as payload to the topic. 
whould be great!

So, in fact this is not a bug, its a feature request.

regards Martin

**Comment by ickerwx** [comment]
I don't have an MQTT server (and whatever else is required to get MQTT up and running) available. Still, it does sound rather easy to implement, you can probably do it yourself and test it in your infrastructure. Googling for python and mqtt turned up a library "paho-mqtt" that looks like it does what you need.

Basically, as a first shot, you'd do something like this:
- create a new proxy module, mqtt.py or whatever
- inside the constructor, create a new mqtt client object and connect to your server, keep a reference to your client as a member of `self` (e.g. `self.mqttclient`)
- inside the `execute()` method, you simply take `data` and use `self.mqttclient.publish("your/topic", codecs.encode(data, 'hex'))` to send data to your server
- once that works, implement some sanity checking on the data if you need it, and create command line parameters for stuff like mqtt server address and credentials, the topic and so on.

I'm not going to setup infrastructure to test this, but I will gladly help you with any problems you come across while implementing it. I'll keep this issue open for a little bit if you want to discuss in here.

**Comment by MatzeJoerling** [comment]
Hi,
am not, really not, a programmer. i hacked the mqtt stuff into the hexlog module.
now it looks like this and is working.
[https://pastebin.com/3kij6pBy](https://pastebin.com/3kij6pBy) (link valid for one week)
(sorry, copy paste python code here didnt work for me)

perhaps you can beautify it and i will test it with parameters.
Clientname, User, Passwort, Server and Topic (envertech/raw) needs to be configured, Port default 1883, username,pass optional

regards Martin

**Comment by ickerwx** [comment]
Alright, if you checkout the `mod_mqtt` branch you will find a new module, `mqtt`.
Use it like so:
```
-om mqtt:server=192.168.x.y:username=solar:password=foobar:topic=envertech/raw:clientid=solar
```
What you did with the `hexdump` module was pretty weird, you were sending the whole output to your server, including offsets and string representation. The output of the new module is a simple hex-encoded string, that means the string `"AAAA\r\n"` will be sent to the server as `"414141410d0a"`.

I did not test this even once, so please run it and let me know how it breaks. I assume it will not work at first, we have to do the debugging in here unless you can fix the errors yourself.

**Comment by MatzeJoerling** [comment]
Hi,

1st, sending the hex stuff that way was the easiest for me to do, i send it to mqtt where i read the data with node-red, i could split sthe tring to an array and work with that.

2nd, more generic would be to send the data as ascii or hex, depends on usage. 

3rd: i got a traceback, but it complains about name resoluition, but there are only IP addresse, i did not get it:

./tcpproxy.py --targetip 47.91.242.120 --targetport 10013 --listenport 10013 -om mqtt:server=192.168.3.48:username=solar:password=xxxxxx:topic=envertech/hex:clientid=solar
Traceback (most recent call last):
  File "./tcpproxy.py", line 418, in <module>
    main()
  File "./tcpproxy.py", line 384, in main
    out_modules = generate_module_list(args.out_modules, incoming=False, verbose=args.verbose)
  File "./tcpproxy.py", line 100, in generate_module_list
    modlist.append(sys.modules['proxymodules.' + name].Module(incoming, verbose, options))
  File "/home/martin/tcpproxy-mod_mqtt/proxymodules/mqtt.py", line 39, in __init__
    self.mqtt.connect(self.server, self.port)
  File "/usr/lib/python3/dist-packages/paho/mqtt/client.py", line 768, in connect
    return self.reconnect()
  File "/usr/lib/python3/dist-packages/paho/mqtt/client.py", line 895, in reconnect
    sock = socket.create_connection((self._host, self._port), source_address=(self._bind_address, 0))
  File "/usr/lib/python3.6/socket.py", line 704, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.6/socket.py", line 745, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Temporary failure in name resolution

regards Martin

**Comment by MatzeJoerling** [comment]
Hi,

when i change connec to: self.mqtt.connect("192.168.3.48", "1883", 60)
i get this one:
Traceback (most recent call last):
  File "./tcpproxy.py", line 418, in <module>
    main()
  File "./tcpproxy.py", line 384, in main
    out_modules = generate_module_list(args.out_modules, incoming=False, verbose=args.verbose)
  File "./tcpproxy.py", line 100, in generate_module_list
    modlist.append(sys.modules['proxymodules.' + name].Module(incoming, verbose, options))
  File "/home/martin/tcpproxy-mod_mqtt/proxymodules/mqtt2.py", line 39, in __init__
    self.mqtt.connect("192.168.3.48", "1883", 60)
  File "/usr/lib/python3/dist-packages/paho/mqtt/client.py", line 767, in connect
    self.connect_async(host, port, keepalive, bind_address)
  File "/usr/lib/python3/dist-packages/paho/mqtt/client.py", line 823, in connect_async
    if port <= 0:
TypeError: '<=' not supported between instances of 'str' and 'int'

M.

**Comment by ickerwx** [comment]
To fix your last error, remove the " around the port number. The port must be a number.

**Comment by MatzeJoerling** [comment]
working with that patch:
39c39
<             self.mqtt.connect(self.server, self.port)
---
>             self.mqtt.connect("192.168.3.34", 1883)

something wrong with self.server
./tcpproxy.py --targetip 47.91.242.120 --targetport 10013 --listenport 10013 -om mqtt:server=192.168.3.34:username=solar:password=xxxxxxx:topic=envertech/hex:clientid=solar

**Comment by ickerwx** [comment]
It was a copy and paste error, check out the latest commit 84bbf34.

**Comment by MatzeJoerling** [comment]
1. error message for mqtt module?? but it was a dependency error for  python3-distutils:

 ./tcpproxy.py --targetip 47.91.242.120 --targetport 10013 --listenport 10013 -om mqtt:server=192.168.3.48:username=solar:password=xxxxxx:topic=envertech/hex:clientid=solar:hex=true
Module mqtt not found

sudo apt install python3-distutils

2,
Exception in thread Thread-11:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "./tcpproxy.py", line 281, in start_proxy_thread
    args.verbose)
  File "./tcpproxy.py", line 180, in handle_data
    data = m.execute(data)
  File "/home/martin/tcpproxy-mod_mqtt/proxymodules/mqtt.py", line 54, in execute
    if not self.mqtt.is_connected():
AttributeError: 'Client' object has no attribute 'is_connected'

**Comment by ickerwx** [comment]
Your first point: it worked before, so it must be a problem on your end. I don't see why installing distutils helped, none of my changes had anything to do with that.

2. I am using the paho mqtt library in version 1.5.0, this is the current release. It supports the `is_connected` function. Check which version you are running and update to the latest release.

**Comment by MatzeJoerling** [comment]
Problem was on my end, distutils was not installed. You use that in 3 modules:
~/tcpproxy-mod_mqtt/proxymodules$ grep dist *
mqtt.py:from distutils.util import strtobool
size404.py:from distutils.util import strtobool
size.py:from distutils.util import strtobool
 didn´t use on of them so no error before. After installing distutils it is working!

ubuntu 18.04 (which i use) has paho 1.3.x , have installed 1.5.0 - working now. 
Will do some tests and report back. 

tnx Martin

**Comment by ickerwx** [comment]
Ah, my bad, I forgot about strtobool.

**Comment by ickerwx** [comment]
So, did you have a chance to test the module? If so, anything that needs fixing?

**Comment by MatzeJoerling** [comment]
Hi,
module is running fine, but i get some "connection refused" from china cloud that results in 
```
47.91.242.120:10013 - Connection refused
47.91.242.120:10013 - Connection refused
Exception in thread Thread-54576:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "./tcpproxy.py", line 270, in start_proxy_thread
    peer = sock.getpeername()
OSError: [Errno 107] Transport endpoint is not connected
```
perhaps you could add a timestamp when this happens or add syslog support (new feature request ;-) )

But, for this issue / feature request i consider the module mqtt working fine.

regards Martin

**Comment by ickerwx** [comment]
Alright, I merged it into master with c3ecf37. Regarding the exception I opened another issue, #15.
